File: Beecrowd_Parte3/O_Problema_da_Parada_1405.java
import java.util.*;
//erro de tudo
public class O_Problema_da_Parada_1405 {
    static final int MAX = 1000;
    static int[] R = new int[10];
    static String[] program;
    static boolean[][] visited;
    static int L;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (true) {
            L = sc.nextInt();
            int N = sc.nextInt();
            if (L == 0 && N == 0) break;
            program = new String[L];
            visited = new boolean[L][MAX];
            for (int i = 0; i < L; i++) {
                program[i] = sc.nextLine().trim();
            }
            Arrays.fill(R, 0);
            R[0] = N;
            int result = run(0);
            if (result == -1) {
                System.out.println("*");
            } else {
                System.out.println(result);
            }
        }
        sc.close();
    }

    static int run(int line) {
        if (line == L) return R[9];
        if (visited[line][R[0]]) return -1;
        visited[line][R[0]] = true;
        String[] parts = program[line].split(" ");
        String cmd = parts[0];
        if (cmd.equals("MOV")) {
            int[] args = parseArgs(parts[1]);
            R[args[0]] = args[1];
            return run(line + 1);
        } else if (cmd.equals("ADD")) {
            int[] args = parseArgs(parts[1]);
            R[args[0]] = (R[args[0]] + args[1]) % MAX;
            return run(line + 1);
        } else if (cmd.equals("SUB")) {
            int[] args = parseArgs(parts[1]);
            R[args[0]] = (R[args[0]] - args[1] + MAX) % MAX;
            return run(line + 1);
        } else if (cmd.equals("MUL")) {
            int[] args = parseArgs(parts[1]);
            R[args[0]] = (R[args[0]] * args[1]) % MAX;
            return run(line + 1);
        } else if (cmd.equals("DIV")) {
            int[] args = parseArgs(parts[1]);
            R[args[0]] = args[1] == 0 ? 0 : R[args[0]] / args[1];
            return run(line + 1);
        } else if (cmd.equals("MOD")) {
            int[] args = parseArgs(parts[1]);
            R[args[0]] = args[1] == 0 ? 0 : R[args[0]] % args[1];
            return run(line + 1);
        } else if (cmd.equals("IFEQ")) {
            int[] args = parseArgs(parts[1]);
            if (R[args[0]] == args[1]) return run(line + 1);
            else return run(findEndif(line) + 1);
        } else if (cmd.equals("IFNEQ")) {
            int[] args = parseArgs(parts[1]);
            if (R[args[0]] != args[1]) return run(line + 1);
            else return run(findEndif(line) + 1);
        } else if (cmd.equals("IFG")) {
            int[] args = parseArgs(parts[1]);
            if (R[args[0]] > args[1]) return run(line + 1);
            else return run(findEndif(line) + 1);
        } else if (cmd.equals("IFL")) {
            int[] args = parseArgs(parts[1]);
            if (R[args[0]] < args[1]) return run(line + 1);
            else return run(findEndif(line) + 1);
        } else if (cmd.equals("IFGE")) {
            int[] args = parseArgs(parts[1]);
            if (R[args[0]] >= args[1]) return run(line + 1);
            else return run(findEndif(line) + 1);
        } else if (cmd.equals("IFLE")) {
            int[] args = parseArgs(parts[1]);
            if (R[args[0]] <= args[1]) return run(line + 1);
            else return run(findEndif(line) + 1);
        } else if (cmd.equals("CALL")) {
            int arg = parseArg(parts[1]);
            int result = run(0);
            if (result == -1) return -1;
            R[9] = result;
            return run(line + 1);
        } else if (cmd.equals("RET")) {
            return parseArg(parts[1]);
        } else {
            return -1;
        }
    }

    static int[] parseArgs(String s) {
        String[] parts = s.split(",");
        return new int[]{parseArg(parts[0]), parseArg(parts[1])};
    }

    static int parseArg(String s) {
        if (s.charAt(0) == 'R') {
            return R[s.charAt(1) - '0'];
        } else {
            return Integer.parseInt(s);
        }
    }

    static int findEndif(int line) {
        int count = 1;
        while (count > 0) {
            line++;
            if (program[line].startsWith("IF")) count++;
            else if (program[line].equals("ENDIF")) count--;
        }
        return line;
    }
}

File: Beecrowd_Parte3/O_Problema_da_Parada_1405.c
#include <stdio.h>
#include <string.h>
//erro de tudo
#define MAX 1000
#define MAX_LINES 100

int R[10];
char program[MAX_LINES][50];
int visited[MAX_LINES][MAX];
int L;

int run(int line);
int parseArg(char* s);
int findEndif(int line);

int main() {
    int N;
    while (scanf("%d %d\n", &L, &N), L || N) {
        for (int i = 0; i < L; i++) {
            fgets(program[i], 50, stdin);
        }
        memset(R, 0, sizeof(R));
        R[0] = N;
        memset(visited, 0, sizeof(visited));
        int result = run(0);
        if (result == -1) {
            printf("*\n");
        } else {
            printf("%d\n", result);
        }
    }
    return 0;
}

int run(int line) {
    if (line == L) return R[9];
    if (visited[line][R[0]]) return -1;
    visited[line][R[0]] = 1;
    char cmd[10];
    char args[20];
    sscanf(program[line], "%s %s", cmd, args);
    if (strcmp(cmd, "MOV") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        R[a] = parseArg(b);
        return run(line + 1);
    } else if (strcmp(cmd, "ADD") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        R[a] = (R[a] + parseArg(b)) % MAX;
        return run(line + 1);
    } else if (strcmp(cmd, "SUB") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        R[a] = (R[a] - parseArg(b) + MAX) % MAX;
        return run(line + 1);
    } else if (strcmp(cmd, "MUL") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        R[a] = (R[a] * parseArg(b)) % MAX;
        return run(line + 1);
    } else if (strcmp(cmd, "DIV") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        R[a] = parseArg(b) == 0 ? 0 : R[a] / parseArg(b);
        return run(line + 1);
    } else if (strcmp(cmd, "MOD") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        R[a] = parseArg(b) == 0 ? 0 : R[a] % parseArg(b);
        return run(line + 1);
    } else if (strcmp(cmd, "IFEQ") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        if (R[a] == parseArg(b)) return run(line + 1);
        else return run(findEndif(line) + 1);
    } else if (strcmp(cmd, "IFNEQ") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        if (R[a] != parseArg(b)) return run(line + 1);
        else return run(findEndif(line) + 1);
    } else if (strcmp(cmd, "IFG") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        if (R[a] > parseArg(b)) return run(line + 1);
        else return run(findEndif(line) + 1);
    } else if (strcmp(cmd, "IFL") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        if (R[a] < parseArg(b)) return run(line + 1);
        else return run(findEndif(line) + 1);
    } else if (strcmp(cmd, "IFGE") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        if (R[a] >= parseArg(b)) return run(line + 1);
        else return run(findEndif(line) + 1);
    } else if (strcmp(cmd, "IFLE") == 0) {
        int a, b;
        sscanf(args, "%d,%d", &a, &b);
        if (R[a] <= parseArg(b)) return run(line + 1);
        else return run(findEndif(line) + 1);
    } else if (strcmp(cmd, "CALL") == 0) {
        int a;
        sscanf(args, "%d", &a);
        int result = run(parseArg(a));
        if (result == -1) return -1;
        R[9] = result;
        return run(line + 1);
    } else if (strcmp(cmd, "RET") == 0) {
        int a;
        sscanf(args, "%d", &a);
        return parseArg(a);
    } else {
        return -1;
    }
}

int parseArg(char* s) {
    if (s[0] == 'R') {
        return R[s[1] - '0'];
    } else {
        return atoi(s);
    }
}

int findEndif(int line) {
    int count = 1;
    while (count > 0) {
        line++;
        if (strncmp(program[line], "IF", 2) == 0) count++;
        else if (strcmp(program[line], "ENDIF") == 0) count--;
    }
    return line;
}